## Challenge 1


### Programming challenge

For this project I have been asked to collect Twitter data using ```twarc``` in order to find out which was the most popular language in the dataset and also to calculate the average, minimum and maximum number of followers that users in the dataset have.

The tasks I had to complete were:

1. Display the first 5 rows
2. Check which columns the dataframe contains
3. Check which columns have missing values
4. Check how many tweets are available in the dataset
5. Calculate the average and the standard deviation of retweets and favorites
6. Indicate the most popular languages in the dataset
7. Calculate the average, minimum and maximum number of followers that users in the dataset have

### Step-by-step Interpretation:

**Pandas library has been imported.**

In [1]:
import pandas as pd

**Importing and naming the Dataset (the Dataset has been downloaded in Python with Twarc via Twitter API).**

In [2]:
df_ev = pd.read_json('Eurovision.jsonl', lines=True)

**Completing first task asking Python to show the first five rows of the Dataset.**

In [3]:
df_ev.head()

created_at                   id               id_str  \
0 2022-05-20 21:25:01+00:00  1527762341298946053  1527762341298946048   
1 2022-05-20 21:25:00+00:00  1527762337503100928  1527762337503100928   
2 2022-05-20 21:24:58+00:00  1527762329789669378  1527762329789669376   
3 2022-05-20 21:24:58+00:00  1527762329168912384  1527762329168912384   
4 2022-05-20 21:24:57+00:00  1527762325284892672  1527762325284892672   

                                           full_text  truncated  \
0  Netherlands so robbed this year, i will never ...      False   
1  RT @Drazmihailovitx: Tanto la legitimación en ...      False   
2  22. Ireland 🇮🇪\n\nbrooke should’ve qualified a...      False   
3  RT @IamLeroySanchez: Making History 🇪🇸 @Chanel...      False   
4  RT @DaniMayakovski: Estos son los heroes occid...      False   

  display_text_range                                           entities  \
0            [0, 88]  {'hashtags': [{'text': 'eurovision', 'indices'...   
1           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2           [0, 109]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3           [0, 119]  {'hashtags': [{'text': 'Eurovision', 'indices'...   
4           [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                            metadata  \
0  {'iso_language_code': 'en', 'result_type': 're...   
1  {'iso_language_code': 'es', 'result_type': 're...   
2  {'iso_language_code': 'en', 'result_type': 're...   
3  {'iso_language_code': 'en', 'result_type': 're...   
4  {'iso_language_code': 'es', 'result_type': 're...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/android" ...           1.527761e+18   
1  <a href="http://twitter.com/download/android" ...                    NaN   
2  <a href="http://twitter.com/download/iphone" r...           1.527761e+18   
3  <a href="http://twitter.com/download/iphone" r...                    NaN   
4  <a href="http://twitter.com/download/android" ...                    NaN   

   ...  favorite_count  favorited  retweeted lang  \
0  ...               0      False      False   en   
1  ...               0      False      False   es   
2  ...               0      False      False   en   
3  ...               0      False      False   en   
4  ...               0      False      False   es   

                                    retweeted_status  \
0                                                NaN   
1  {'created_at': 'Fri May 20 17:16:34 +0000 2022...   
2                                                NaN   
3  {'created_at': 'Fri May 20 11:00:58 +0000 2022...   
4  {'created_at': 'Fri May 20 19:54:20 +0000 2022...   

                                   extended_entities  possibly_sensitive  \
0                                                NaN                 NaN   
1                                                NaN                 NaN   
2  {'media': [{'id': 1524732896736690177, 'id_str...                 0.0   
3  {'media': [{'id': 1527605171995418624, 'id_str...                 0.0   
4                                                NaN                 NaN   

  quoted_status_id  quoted_status_id_str  quoted_status  
0              NaN                   NaN            NaN  
1              NaN                   NaN            NaN  
2              NaN                   NaN            NaN  
3              NaN                   NaN            NaN  
4              NaN                   NaN            NaN  

[5 rows x 31 columns]

**Checking how big is the Dataset (number of rows included). The Twarc query has been kept active not for so long so not a big dataset is expected.**

In [4]:
len(df_ev)

300

**Checking the Columns present in the Dataset and completing task number 2.**

In [5]:
df_ev.columns

Index(['created_at', 'id', 'id_str', 'full_text', 'truncated',
       'display_text_range', 'entities', 'metadata', 'source',
       'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str',
       'in_reply_to_screen_name', 'user', 'geo', 'coordinates', 'place',
       'contributors', 'is_quote_status', 'retweet_count', 'favorite_count',
       'favorited', 'retweeted', 'lang', 'retweeted_status',
       'extended_entities', 'possibly_sensitive', 'quoted_status_id',
       'quoted_status_id_str', 'quoted_status'],
      dtype='object')

**Looking if there are any missing values and completing task number 3.**

In [6]:
df_ev.isna().sum()

created_at                     0
id                             0
id_str                         0
full_text                      0
truncated                      0
display_text_range             0
entities                       0
metadata                       0
source                         0
in_reply_to_status_id        255
in_reply_to_status_id_str    255
in_reply_to_user_id          250
in_reply_to_user_id_str      250
in_reply_to_screen_name      250
user                           0
geo                          300
coordinates                  300
place                        294
contributors                 300
is_quote_status                0
retweet_count                  0
favorite_count                 0
favorited                      0
retweeted                      0
lang                           0
retweeted_status             121
extended_entities            237
possibly_sensitive           218
quoted_status_id             291
quoted_status_id_str         291
quoted_sta

**Using the describe function in order to check how many tweets are available in this Dataset (even if we know that already from the "Len" function which told us already there are 300 tweets).**

In [7]:
df_ev.describe()

id        id_str  in_reply_to_status_id  \
count  3.000000e+02  3.000000e+02           4.500000e+01   
mean   1.527761e+18  1.527761e+18           1.527709e+18   
std    7.469430e+11  7.469430e+11           1.229518e+14   
min    1.527760e+18  1.527760e+18           1.527310e+18   
25%    1.527760e+18  1.527760e+18           1.527742e+18   
50%    1.527761e+18  1.527761e+18           1.527756e+18   
75%    1.527762e+18  1.527762e+18           1.527760e+18   
max    1.527762e+18  1.527762e+18           1.527762e+18   

       in_reply_to_status_id_str  in_reply_to_user_id  \
count               4.500000e+01         5.000000e+01   
mean                1.527709e+18         5.471121e+17   
std                 1.229518e+14         6.634163e+17   
min                 1.527310e+18         1.443632e+07   
25%                 1.527742e+18         2.885561e+08   
50%                 1.527756e+18         1.988654e+09   
75%                 1.527760e+18         1.311599e+18   
max                 1.527762e+18         1.523383e+18   

       in_reply_to_user_id_str  geo  coordinates  contributors  retweet_count  \
count             5.000000e+01  0.0          0.0           0.0     300.000000   
mean              5.471121e+17  NaN          NaN           NaN     131.663333   
std               6.634163e+17  NaN          NaN           NaN     401.379173   
min               1.443632e+07  NaN          NaN           NaN       0.000000   
25%               2.885561e+08  NaN          NaN           NaN       0.000000   
50%               1.988654e+09  NaN          NaN           NaN       7.000000   
75%               1.311599e+18  NaN          NaN           NaN      63.000000   
max               1.523383e+18  NaN          NaN           NaN    4434.000000   

       favorite_count  possibly_sensitive  quoted_status_id  \
count      300.000000           82.000000      9.000000e+00   
mean         0.333333            0.048780      1.527093e+18   
std          1.646590            0.216734      8.498380e+14   
min          0.000000            0.000000      1.525124e+18   
25%          0.000000            0.000000      1.527271e+18   
50%          0.000000            0.000000      1.527474e+18   
75%          0.000000            0.000000      1.527596e+18   
max         16.000000            1.000000      1.527618e+18   

       quoted_status_id_str  
count          9.000000e+00  
mean           1.527093e+18  
std            8.498380e+14  
min            1.525124e+18  
25%            1.527271e+18  
50%            1.527474e+18  
75%            1.527596e+18  
max            1.527618e+18

**As we can see Python is not able to show us the data properly (There are most likely too many zeroes). This command below is needed to set the options for pandas to show floats with a maximum of 5 decimals (that's what the 5 in the '%.5f' does).**

In [8]:
pd.set_option('display.float_format', lambda x: '%.5f' % x)

**Now the numbers are shown properly and as we can see we have 300 Tweets. Task 4 has been completed.**

In [9]:
df_ev.describe()

id                    id_str  \
count                 300.00000                 300.00000   
mean  1527761069162088448.00000 1527761069162088448.00000   
std          746943003891.76855        746943003891.76855   
min   1527759684140158976.00000 1527759684140158976.00000   
25%   1527760487698619392.00000 1527760487698619392.00000   
50%   1527761082911182848.00000 1527761082911182848.00000   
75%   1527761718280115200.00000 1527761718280115200.00000   
max   1527762341298946048.00000 1527762341298946048.00000   

          in_reply_to_status_id  in_reply_to_status_id_str  \
count                  45.00000                   45.00000   
mean  1527708638157762304.00000  1527708638157762304.00000   
std       122951801974023.85938      122951801974023.85938   
min   1527309871246323712.00000  1527309871246323712.00000   
25%   1527741527761485824.00000  1527741527761485824.00000   
50%   1527756125260681216.00000  1527756125260681216.00000   
75%   1527760071756767232.00000  1527760071756767232.00000   
max   1527761951144742912.00000  1527761951144742912.00000   

            in_reply_to_user_id   in_reply_to_user_id_str     geo  \
count                  50.00000                  50.00000 0.00000   
mean   547112112206716928.00000  547112112206716928.00000     NaN   
std    663416323627907968.00000  663416323627907968.00000     NaN   
min              14436317.00000            14436317.00000     NaN   
25%             288556095.50000           288556095.50000     NaN   
50%            1988653961.00000          1988653961.00000     NaN   
75%   1311598573214142464.00000 1311598573214142464.00000     NaN   
max   1523383090189635584.00000 1523383090189635584.00000     NaN   

       coordinates  contributors  retweet_count  favorite_count  \
count      0.00000       0.00000      300.00000       300.00000   
mean           NaN           NaN      131.66333         0.33333   
std            NaN           NaN      401.37917         1.64659   
min            NaN           NaN        0.00000         0.00000   
25%            NaN           NaN        0.00000         0.00000   
50%            NaN           NaN        7.00000         0.00000   
75%            NaN           NaN       63.00000         0.00000   
max            NaN           NaN     4434.00000        16.00000   

       possibly_sensitive          quoted_status_id      quoted_status_id_str  
count            82.00000                   9.00000                   9.00000  
mean              0.04878 1527093130574570752.00000 1527093130574570752.00000  
std               0.21673     849837971808191.37500     849837971808191.37500  
min               0.00000 1525123590932373504.00000 1525123590932373504.00000  
25%               0.00000 1527270783168303104.00000 1527270783168303104.00000  
50%               0.00000 1527473685124218880.00000 1527473685124218880.00000  
75%               0.00000 1527595584819015680.00000 1527595584819015680.00000  
max               1.00000 1527618432820252672.00000 1527618432820252672.00000

**Calculating the average (mean) and the standard deviation of retweets and favorites. Task 5 completed.**

In [10]:
df_ev[['retweet_count', 'favorite_count']].describe()

retweet_count  favorite_count
count      300.00000       300.00000
mean       131.66333         0.33333
std        401.37917         1.64659
min          0.00000         0.00000
25%          0.00000         0.00000
50%          7.00000         0.00000
75%         63.00000         0.00000
max       4434.00000        16.00000

**Checking the language in which evry single Tweet has been wrote down. There were 13 languages present in the Dataset with the majority of them wrote in English or Spanish. Task 6 has been completed.**

In [11]:
df_ev['lang'].value_counts()

en     137
es     117
it      19
und      8
fr       8
uk       3
tr       2
nl       1
sv       1
pt       1
de       1
da       1
el       1
Name: lang, dtype: int64

**Another (faster but less specific) way to check how many languages were present in the Dataset. Among the 300 values there were 13 languages present. The majority of them were written in English and the highest frequency was a value of 137 tweets wrote in the same language.**

In [12]:
df_ev[['lang']].describe()

lang
count   300
unique   13
top      en
freq    137

**Checking the "User" column which has we can see it contains another dictionary in it.**

In [13]:
df_ev['user']

0      {'id': 1126212265869443072, 'id_str': '1126212...
1      {'id': 309612124, 'id_str': '309612124', 'name...
2      {'id': 1048624695451570177, 'id_str': '1048624...
3      {'id': 2531551437, 'id_str': '2531551437', 'na...
4      {'id': 1488398972947517443, 'id_str': '1488398...
                             ...                        
295    {'id': 1328996764511301632, 'id_str': '1328996...
296    {'id': 1518923968639610882, 'id_str': '1518923...
297    {'id': 4409501567, 'id_str': '4409501567', 'na...
298    {'id': 561512960, 'id_str': '561512960', 'name...
299    {'id': 1261795983840681987, 'id_str': '1261795...
Name: user, Length: 300, dtype: object

**Using this function in order to be able to read the dictionary inside the column.**

In [14]:
def expand_dictionary(row, columnname=''):
    if columnname == '':
        return row
    # Returning the original row if the column does not contain a dictionary
    if type(row[columnname]) != dict:
        return row
    
    # Otherwise, transforming the dictionary in additional columns
    for key, value in row[columnname].items():
        if type(value) != dict:
            row[key] = value
        else:
            for subkey, subvalue in row[columnname][key].items():
                if type(subvalue) != dict:
                    row[str(key)+'_'+str(subkey)] = subvalue
                else:
                    for subsubkey, subsubvalue in row[columnname][key][subkey].items():
                        row[str(key)+'_'+str(subkey)+'_'+str(subsubkey)] = subsubvalue
                
        
    return row 

**Creating a new dataset which contains the additional dictionary present in the "User" columns.**

In [15]:
df_ev2 = df_ev.apply(expand_dictionary, axis=1, args=('user',))

**Checking which additional columns the new dictionary was containing.**

In [16]:
df_ev2.columns

Index(['contributors', 'contributors_enabled', 'coordinates', 'created_at',
       'default_profile', 'default_profile_image', 'description',
       'display_text_range', 'entities', 'entities_description_urls',
       'entities_url_urls', 'extended_entities', 'favorite_count', 'favorited',
       'favourites_count', 'follow_request_sent', 'followers_count',
       'following', 'friends_count', 'full_text', 'geo', 'geo_enabled',
       'has_extended_profile', 'id', 'id_str', 'in_reply_to_screen_name',
       'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str', 'is_quote_status',
       'is_translation_enabled', 'is_translator', 'lang', 'listed_count',
       'location', 'metadata', 'name', 'notifications', 'place',
       'possibly_sensitive', 'profile_background_color',
       'profile_background_image_url', 'profile_background_image_url_https',
       'profile_background_tile', 'profile_banner_url', 'profile_image_url',
      

**After finding the columns we needed, "followers_count", we can check the minimum, the average and the max followers the Users of this Dataset have aggregating this values together.**

In [17]:
df_ev2['followers_count'].agg(['mean','min','max'])

mean     29538.33333
min          0.00000
max    8557449.00000
Name: followers_count, dtype: float64

**Otherwise we can run a simple command as such the one below and see that the average number of followers of our users is 29,5K, the minumum is 0 and the max is 8,5 Million followers. Task 7 completed.**

In [18]:
df_ev2['followers_count'].describe()

count       300.00000
mean      29538.33333
std      494014.47779
min           0.00000
25%          86.75000
50%         322.50000
75%        1065.00000
max     8557449.00000
Name: followers_count, dtype: float64

## Challenge 2

### Programming challenge

For this project I have been asked to run a sentiment analysis on social media data (Twitter) in order to predict approximately how many followers of the company will be willing to participate to a fundraising campaign run on social media and effectively make a donation.

Since the sentiment analysys will be made via Azure only on Tweets written in English I first need to clean and slice the Dataset of tweets collected via the Twitter API.

I will need need to:
1. Select only tweets that have language set to English (i.e., ```en```)
2. Select only the columns ```id``` and ```text```

After slicing the dataframe following the needs above I need to:
1. Save the results to a pickle file to request a sentyment analysis.

### Step-by-step Interpretation:

**Importing Pandas**

In [19]:
import pandas as pd

**Importing the dataset.**

In [20]:
df_jsonl = pd.read_json('greenpeacetweets.jsonl', lines=True)

**In order to ask Python to visualize only the tweets in english I have to ask it to verify which of the tweets are in english language. With this operation Pandas will verify this row by row.**

In [21]:
df_jsonl['lang']=='en'

0       False
1       False
2       False
3        True
4        True
        ...  
1192    False
1193    False
1194    False
1195    False
1196    False
Name: lang, Length: 1197, dtype: bool

**Now I am asking Python to display these Tweets, slicing the Dataset by rows.**

In [22]:
df_jsonl[df_jsonl['lang']=='en']


created_at                   id               id_str  \
3    2021-09-15 20:21:02+00:00  1438236436084391936  1438236436084391936   
4    2021-09-15 20:20:54+00:00  1438236402114772995  1438236402114772992   
6    2021-09-15 20:20:23+00:00  1438236275283214343  1438236275283214336   
9    2021-09-15 20:18:42+00:00  1438235849058045956  1438235849058045952   
10   2021-09-15 20:18:38+00:00  1438235834415714308  1438235834415714304   
...                        ...                  ...                  ...   
1167 2021-09-15 17:26:36+00:00  1438192539400429577  1438192539400429568   
1168 2021-09-15 17:26:34+00:00  1438192531787825157  1438192531787825152   
1177 2021-09-15 17:24:59+00:00  1438192131206782979  1438192131206782976   
1183 2021-09-15 17:24:32+00:00  1438192017847226369  1438192017847226368   
1186 2021-09-15 17:24:29+00:00  1438192006350651396  1438192006350651392   

                                              full_text  truncated  \
3     RT @PBIcanada: We congratulate @GreenpeaceCA o...      False   
4     RT @cadebe_: @GJGamble @BorisJohnson @trussliz...      False   
6     RT @GreenpeaceCA: I scream, you scream, WE ALL...      False   
9     RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...      False   
10    RT @zoev213: Happy 50th @greenpeaceusa Join me...      False   
...                                                 ...        ...   
1167  @cowboy_satin @23brookside @People4Bernie @Ber...      False   
1168  RT @jubileam8: #Glyphosate spraying purposeful...      False   
1177  RT @EDProgramme: Companies like ExxonMobil kne...      False   
1183  RT @AbukaAlfred: Different countries, differen...      False   
1186    happy 50th Greenpeace ! https://t.co/gtBWdYOGRI      False   

     display_text_range                                           entities  \
3              [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4              [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   
6              [0, 140]  {'hashtags': [{'text': 'TAXTHERICH', 'indices'...   
9              [0, 140]  {'hashtags': [], 'symbols': [], 'user_mentions...   
10             [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
...                 ...                                                ...   
1167         [106, 192]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1168           [0, 140]  {'hashtags': [{'text': 'Glyphosate', 'indices'...   
1177           [0, 140]  {'hashtags': [{'text': 'ClimateChange', 'indic...   
1183           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1186            [0, 23]  {'hashtags': [], 'symbols': [], 'user_mentions...   

     extended_entities                                           metadata  \
3                  NaN  {'iso_language_code': 'en', 'result_type': 're...   
4                  NaN  {'iso_language_code': 'en', 'result_type': 're...   
6                  NaN  {'iso_language_code': 'en', 'result_type': 're...   
9                  NaN  {'iso_language_code': 'en', 'result_type': 're...   
10                 NaN  {'iso_language_code': 'en', 'result_type': 're...   
...                ...                                                ...   
1167               NaN  {'iso_language_code': 'en', 'result_type': 're...   
1168               NaN  {'iso_language_code': 'en', 'result_type': 're...   
1177               NaN  {'iso_language_code': 'en', 'result_type': 're...   
1183               NaN  {'iso_language_code': 'en', 'result_type': 're...   
1186               NaN  {'iso_language_code': 'en', 'result_type': 're...   

                                                 source  ...  is_quote_status  \
3     <a href="http://twitter.com/download/iphone" r...  ...            False   
4     <a href="http://twitter.com/download/android" ...  ...            False   
6     <a href="https://mobile.twitter.com" rel="nofo...  ...            False   
9     <a href="http://twitter.com/download/iphone" r...  ...            False   

**Now I am asking Python to display only the columns of "id" and "full text" of the Tweets, slicing the Dataset by columns.**

In [23]:
df_jsonl[['id','full_text']]

id                                          full_text
0     1438236491394785288  RT @GreenpeaceArg: - ¿Por qué tan elegante, Ho...
1     1438236482532151301  ¡Excusas, excusas, excusas! Cómo las empresas ...
2     1438236443638370308  RT @Herbert_Diess: Greenpeace müsste man erfin...
3     1438236436084391936  RT @PBIcanada: We congratulate @GreenpeaceCA o...
4     1438236402114772995  RT @cadebe_: @GJGamble @BorisJohnson @trussliz...
...                   ...                                                ...
1192  1438191764624580608  @lgewessler @NicolaWerdenigg @Greenpeace leben...
1193  1438191764125462535  RT @nickofnz: It’s time to respect the rāhui, ...
1194  1438191713512828939  RT @schnellenbachj: Toll, wenn man eine von de...
1195  1438191669514485773  RT @G_Simonetti: Hoy Greenpeace cumple 50 años...
1196  1438191603454255106  RT @GreenpeaceArg: - ¿Por qué tan elegante, Ho...

[1197 rows x 2 columns]

**Creating a new Dataset containing only a portion (only English Tweets) of the original Dataset (I do that in order to keep the original Dataset always untouched in case I need to go back for some mistake or corrections)**

In [24]:
en_tweet = df_jsonl[df_jsonl['lang']=='en']

**Slicing the new Dataset by column. Only tweets in English are being displayed and only the "id" and "full text" columns. Task 1 and 2 completed.**

In [25]:
en_tweet[['id','full_text']]

id                                          full_text
3     1438236436084391936  RT @PBIcanada: We congratulate @GreenpeaceCA o...
4     1438236402114772995  RT @cadebe_: @GJGamble @BorisJohnson @trussliz...
6     1438236275283214343  RT @GreenpeaceCA: I scream, you scream, WE ALL...
9     1438235849058045956  RT @LongTimeAmy: TODAY IN HISTORY\n\n1971–Envi...
10    1438235834415714308  RT @zoev213: Happy 50th @greenpeaceusa Join me...
...                   ...                                                ...
1167  1438192539400429577  @cowboy_satin @23brookside @People4Bernie @Ber...
1168  1438192531787825157  RT @jubileam8: #Glyphosate spraying purposeful...
1177  1438192131206782979  RT @EDProgramme: Companies like ExxonMobil kne...
1183  1438192017847226369  RT @AbukaAlfred: Different countries, differen...
1186  1438192006350651396    happy 50th Greenpeace ! https://t.co/gtBWdYOGRI

[420 rows x 2 columns]